feat(core): add Internationalization to Catalyst

bc-alexsaiannyi opened the pull request:
## What/Why?
This PR adds i18n mechanism to Catalyst. It's basically just a mechanism and will be expanded to all Pages & Components in upcoming PRs.

## Ticket
[CATALYST-266](https://bigcommercecloud.atlassian.net/browse/CATALYST-266)

## Testing
locally

https://github.com/bigcommerce/catalyst/assets/67792608/abb4b1d0-afd2-4df0-b863-8cdca0494a09



[CATALYST-266]: https://bigcommercecloud.atlassian.net/browse/CATALYST-266?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

BC-krasnoshapka commented:
An update to someone who will be reviewing this PR. 

As Alex wrote before, locale prefixes in URLs are temporary disabled until we have full picture of multi-language implementation on headless stores and default approach for internationalized routes. So this is not final implementation and we'll layer on top of it later.

We discussed removing `[locale]` dynamic parameter from the app router, but seems it's not possible as this is the only way how `next-intl` library passes current locale through component tree - via middleware and locale prop. We could hack things here and there, but that goes against library ideology and probably will be unstable.

christensenep commented on apps/core/app/[locale]/(default)/(faceted)/brand/[slug]/static/page.tsx:
```diff
@@ -6,9 +8,11 @@ export default BrandPage;
 
 export async function generateStaticParams() {
   const brands = await getBrands();
+  const locale = await getLocale();
```
How is this working with static generated pages? Is it generating a static page for every supported locale? That doesn't *look* like what's happening based on what I'm reading here, but could be misunderstanding.

bc-alexsaiannyi replied:
You're right. I've moved it to Home page. Guess, it's a right place.

christensenep commented on apps/core/app/[locale]/(default)/product/[slug]/page.tsx:
```diff
@@ -235,6 +239,7 @@ export default async function Product({ params, searchParams }: ProductPageProps
 
   return (
     <>
+      <p>{t('test')}</p>
```
Make sure to remove

bc-alexsaiannyi commented:
**NOTE:** 
Playwright UI Tests are failing due to changes in locators text. 
For example,
**before** `waiting for getByText('1 items')`  
**now** it would look like - `waiting for getByText('(EN )1 item')`

Since phrases in dictionaries are temporal and will be modified later, I think we can skip it temporary

jorgemoya reviewed (commented):
Seems to be deleting pages that should not be deleted, unless I'm missing something?

Edit: nevermind, I see now what happened.

jorgemoya reviewed (commented):
Hey @bc-alexsaiannyi. I believe this could be due to not reviewing this PR sooner (sorry 🙏 ), but the length of this PR is too big to feel 100% comfortable nothing will break (or some new change won't be reverted). 

Would it be possible to split this PR into several PRs? Maybe a setup PR that adds all the logic we need and the multiple PRs to translate routes?

What do you think?

bc-alexsaiannyi commented:
> Hey @bc-alexsaiannyi. I believe this could be due to not reviewing this PR sooner (sorry 🙏 ), but the length of this PR is too big to feel 100% comfortable nothing will break (or some new change won't be reverted).
>  ...
> What do you think?

Hey @jorgemoya. First of all, thank you so much for reviewing all this! I do aware that such volume of changes are definitely massive in terms of review.  But for passed time w've got a history of changes and discussions related to i18n and it can be helpful for understanding of what’ve been done. 

I hope that going through commits may do a review little bit easier🙏 and I can name several commits that basically provide main logic and are crucial in terms of internationalization mechanism, while others (as you’ve noticed)  just bring translations for pages & components.

So what do you think about reviewing next commits?🤔

1. feat(core): add Internationalization mechanism - e734c3509dacaded58c18623baac279fcc79bad0
2. feat(core): simplified middlewares, mostly rely on locale detection… -   492afe625f409733eb034fcafa5781944c88bda2
3. feat(core): fixed case when cookie is empty on first load … - bbf872ccb22224e7f0c536fd3703866930ae3c12

jorgemoya commented on apps/core/app/[locale]/not-found.tsx:
```diff
@@ -30,23 +37,29 @@ export default async function NotFound() {
+                messages={pick(dictionary, 'Product')}
+              >
+                <ProductCard
+                  key={product.entityId}
```
I don't think we need a key here anymore if inside Provider.

bc-alexsaiannyi replied:
Yeah, i've missed it here.

jorgemoya commented on apps/core/app/[locale]/(default)/(faceted)/brand/[slug]/page.tsx:
```diff
@@ -0,0 +1,139 @@
+export default async function Brand({ params: { slug, locale }, searchParams }: Props) {
+  unstable_setRequestLocale(locale);
+
+  const t = await getTranslations({ locale, namespace: 'Brand' });
```
Why do we need to pass the locale here (and other similar instances)?

jorgemoya reviewed (commented):
Looking good just some comments I have since I'm not very familiar with `next-intl`.

bc-alexsaiannyi commented on apps/core/i18n.ts:
```diff
@@ -0,0 +1,33 @@
+
+type LocalePrefixesType = `${LocalePrefixes}`;
+
+// Temporary we use NEVER prefix to prioritize accept-language header
+// & disable internationalized routes due to multi-language in Catalyst is in progress
+const localePrefix: LocalePrefixesType = LocalePrefixes.NEVER;
```
During request `accept-language` header just helps to detect preferred language & locale for a user but in terms how next-intl handles routing it stays in 3rd place in its list of [priorities](https://next-intl-docs.vercel.app/docs/routing/middleware#prefix-locale-detection):
1. A locale prefix is present in the pathname (e.g. /en/page)
2. A cookie is present that contains a previously detected locale
3. A locale can be matched based on the accept-language header

Most libraries work with internationalized routes out of the box and that’s what we want to get at the end but as Nathan mentioned [here](https://github.com/bigcommerce/catalyst/pull/463#issuecomment-1919780600) not everything can be localised at the moment. And while we’re improving things we can temporary switch our strategy and shift priority from internationalized routes to accept-lang header that is kind of what we’re doing in corner as well. 

Since library can handle lang header we’ve just decided to switch off usage of prefixes in routes to raise up header’s priority so now when a User comes to a store at first time, library takes locale/lang preferences from the header & saves its to cookies and re-uses it later on.

And once we want to change the behaviour, we always can turn back to prefixes in routing with one “click”.

bc-alexsaiannyi commented on apps/core/middlewares/with-routes.ts:
```diff
@@ -24,6 +26,25 @@ interface StorefrontStatusCache {
 }
 
+let locale: string;
+const clearPathFromLocale = (path: string) => {
```
Yeah, good catch. I guess it sounds better)

bc-alexsaiannyi commented on apps/core/dictionaries/en.json:
```diff
@@ -0,0 +1,229 @@
+{
+  "RootLayout": {
+    "Metadata": {
+      "title": "(EN) Catalyst Store",
```
well, quite frankly, we don’t need them anywhere. It’s been just a way to demonstrate that we switch between languages plus it helps to control coverage of all static phrases in the project. I’m opened to suggestions and think that we can remove this “prefix” from dictionaries if we’re agreed on that. What do you say?

jorgemoya replied:
I guess this is a question for @bookernath. I'd prefer if english has no prefixes and only add to other languages if they're not translated.

bc-alexsaiannyi commented on apps/core/components/nav-link/index.tsx:
```diff
@@ -0,0 +1,34 @@
+'use client';
+
+import { ComponentProps, ElementRef, forwardRef } from 'react';
+
+import { cn } from '~/lib/utils';
+
+import { Link } from '../../navigation';
+
+type NavLinkProp = ComponentProps<typeof Link>;
+type NavLinkType = Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, keyof NavLinkProp> &
+  NavLinkProp & {
+    children?: React.ReactNode;
+  } & React.RefAttributes<HTMLAnchorElement>;
+
+// This component wraps next/link and automatically prefixes the href with the current locale as necessary.
+// https://next-intl-docs.vercel.app/docs/routing/navigation#link
+export const NavLink = forwardRef<ElementRef<'a'>, NavLinkType>(
+  ({ href, prefetch = false, children, className, ...rest }, ref) => {
+    return (
+      <Link
+        className={cn(
+          ' hover:text-blue-primary focus:outline-none focus:ring-4 focus:ring-blue-primary/20',
+          className,
+        )}
+        href={href}
+        prefetch={prefetch}
+        ref={ref}
+        {...rest}
+      >
+        {children}
+      </Link>
+    );
+  },
+);
```
yes, I’ve seen it earlier  but believe that we are still better to use Link from the library since it has extended [functionality](https://next-intl-docs.vercel.app/docs/routing/navigation#link) for localisation. 

For now we’ve made a configuration `Navigation` in the same named file with [shared pathnames](https://next-intl-docs.vercel.app/docs/routing/navigation#shared-pathnames) and library provides utils for this purpose. Later, if we or catalyst’s users would like to switch to [localised pathnames](https://next-intl-docs.vercel.app/docs/routing/navigation#localized-pathnames) it’s still covered with library but in both cases they provide us their version of wrapper for `next/link` component. 

I’m thinking that maybe we can bring here( to nav-link) and kind of merge what’ve been done in `components/link` to sync them. What do you think about this?

jorgemoya replied:
I think we just need to keep `Link` and instead of using `NextLink` use `Link` from library, but maybe @deini has a better approach.

bc-alexsaiannyi commented on apps/core/components/nav-link/index.tsx:
```diff
@@ -0,0 +1,78 @@
+'use client';
```
Yeah, it's what I've thought before. Just didn't want to add changes before we agree on that. I'Il do it now.
